### ISSUE TITLE
Which would be your preferred back-end technology/stack/framework for a React web app?

### ISSUE THREAD

**Issue by helloguille** [mxstbr/ama#51]

**Comment by mxstbr** [comment]
The answer to this question depends largely on your use case, requirements and team. Choosing something that your team is familiar with over a new hot technology is a good idea. Go boring rather than exciting, because if you're choosing the new hot thing that might not be so hot anymore by the time you have to hire people to maintain your system.

That being said, for our newest project I chose to go with a Node.js server running express with a GraphQL API powered by RethinkDB. Yep, GraphQL and RethinkDB. That kind of contradicts what I just said, but let me explain my reasoning:

- **Node.js**: The whole team is JavaScript developers, so that's a no-brainer.
- **Express.js**: It's the standard and has a big ecosystem of tools that we leverage to get stuff done quicker and better. 
- **GraphQL**: The whole team is mainly frontend developers with only a bit of backend experience. REST APIs are pretty frustrating to build as somebody who doesn't do it very often. There are no hard rules on how to do it, which means you can (and I will) fuck it up pretty badly and have to live with it forever. With GraphQL you don't have to think about any of that, you just build the API and because it's so strict it'll probably be the right way of doing it. (or if it isn't, it's easily fixed) It's also amazing to work with from the frontend, the declarative data fetching is a beautiful way of working and ties amazingly into React.
- **RethinkDB**: The standard choice here would've been PostgreSQL with a Redis caching layer. The thing is that none of us know SQL, and the app is realtime. RethinkDB has a beautiful JavaScript API that's easy to understand and query, is made for realtime querying and scales nicely no matter how many people are subscribed. (we might need to add a Redis caching layer later on anyway, but for now it's fine)

Again, those technologies were chosen with _our specific use case in mind_. I'm not saying you should choose these exact tools for your project. (in fact I'm pretty sure that combination is not the best for 99% of people)

If you don't remember anything of this answer then remember this: **Think about your team, requirements and use case, and then come up with a set of tools that'll help you get it done quicker and better**.

**Comment by helloguille** [comment]
Thank you so much for you detailed answer. I'll post my comments next week, but I wanted to thank you in advance.

**Comment by helloguille** [comment]
It is so fascinating to see how younger programmers are approaching web programming today. I started 17 years ago with ASP classic and Java, I did a lot of Javascript but with IE 5 & Co. you had to forget about doing anything serious with it, other than hacking with JQuery and Prototype.

Later on I switched to MVC frameworks (Rails), and even implemented my own MVC in PHP (Something I'm not particularly proud of). And went back into Javascript recently.

Things change so fast, that I had to learn to keep my mind open. I just cannot imagine you'd implement a backend without any SQL right now, but I'm sure there must be a good reasoning behind, but I wont understand it until I dive into it.
